google/cloud/spanner/internal/connection_impl.cc

```diff
@@ -379,90 +416,139 @@ QueryResult ConnectionImpl::ExecuteQueryImpl(
-      return QueryResult(
-          google::cloud::internal::make_unique<StatusOnlyResultSetSource>(
-              Status(StatusCode::kInternal,
-                     "Begin transaction requested but no transaction returned "
-                     "(in ExecuteQuery).")));
+      return T(google::cloud::internal::make_unique<StatusOnlyResultSetSource>(
+          Status(StatusCode::kInternal, begin_transaction_error_message)));
     }
     s.set_id(metadata->transaction().id());
   }
-  return QueryResult(std::move(*reader));
+  return T(std::move(*reader));
 }
 
-StatusOr<DmlResult> ConnectionImpl::ExecuteDmlImpl(
+template <typename T>
+T ConnectionImpl::CommonQueryImpl(
+    SessionHolder& session, spanner_proto::TransactionSelector& s,
+    std::int64_t seqno, ExecuteSqlParams esp,
+    google::spanner::v1::ExecuteSqlRequest::QueryMode query_mode,
+    std::string const& begin_transaction_error_message) {
+  auto const& stub = stub_;
+  auto retry_resume_fn =
+      [stub, this](spanner_proto::ExecuteSqlRequest& request) mutable
+      -> StatusOr<std::unique_ptr<ResultSourceInterface>> {
+    // Capture a copy of `stub` to ensure the `shared_ptr<>` remains valid
+    // through the lifetime of the lambda. Note that the local variable `stub`
+    // is a reference to avoid increasing refcounts twice, but the capture is by
+    // value.
+    auto factory = [stub, request](std::string const& resume_token) mutable {
+      request.set_resume_token(resume_token);
+      auto context =
+          google::cloud::internal::make_unique<grpc::ClientContext>();
+      return google::cloud::internal::make_unique<
+          DefaultPartialResultSetReader>(
+          std::move(context), stub->ExecuteStreamingSql(*context, request));
+    };
+    auto rpc = google::cloud::internal::make_unique<PartialResultSetResume>(
+        std::move(factory), Idempotency::kIdempotent,
+        this->retry_policy_->clone(), this->backoff_policy_->clone());
+
+    return PartialResultSetSource::Create(std::move(rpc));
+  };
+  auto ret_val = ExecuteSqlImpl<T>(session, s, seqno, std::move(esp),
+                                   query_mode, std::move(retry_resume_fn),
+                                   begin_transaction_error_message);
+  if (!ret_val) {
+    return T(google::cloud::internal::make_unique<StatusOnlyResultSetSource>(
+        std::move(ret_val).status()));
+  }
+  return std::move(*ret_val);
+}
+
+QueryResult ConnectionImpl::ExecuteQueryImpl(
     SessionHolder& session, spanner_proto::TransactionSelector& s,
     std::int64_t seqno, ExecuteSqlParams esp) {
-  if (!session) {
-    auto session_or = AllocateSession();
-    if (!session_or) {
-      return std::move(session_or).status();
+  return CommonQueryImpl<QueryResult>(
+      session, s, seqno, std::move(esp),
+      spanner_proto::ExecuteSqlRequest::NORMAL,
+      "Begin transaction requested but no transaction returned "
+      "(in ExecuteQuery).");
```

Comment:
This "in ExecuteQuery" parenthetical is the only place where this error message differs. Is this difference important? If not, we could avoid passing this error string, which would nice.

Reply:
I'm not sure if the difference is important; it does add a small amount of useful diagnostic information.

Reply:
Asking callers to construct a potential error message for an operation that the caller does not call does not feel right. I think we should simply remove this argument from the function. I suspect the extra information will not be critical. If it ever is needed, the called function can compute the string itself based on the template parameter `T` that's specified.